relay-test/src/mini_sentry.rs

```diff
@@ -0,0 +1,570 @@
+use std::collections::{BTreeMap, HashMap};
+use std::future::Future;
+use std::io::Read;
+use std::net::SocketAddr;
+use std::pin::Pin;
+use std::sync::{Arc, Mutex, PoisonError};
+use std::time::Duration;
+
+use axum::body::Bytes;
+use axum::response::Json;
+use axum::routing::{get, post};
+use axum::Router;
+use flate2::read::GzDecoder;
+use serde_json::{json, Value};
+use tokio::runtime::Runtime;
+use uuid::Uuid;
+
+use relay_auth::{PublicKey, RelayId, RelayVersion};
+use relay_base_schema::project::{ProjectId, ProjectKey};
+use relay_config::RelayInfo;
+use relay_dynamic_config::GlobalConfig;
+use relay_event_schema::protocol::EventId;
+
+use crate::{random_port, Envelope, Outcome, ProjectState, RawItem, Upstream};
+
+pub const DEFAULT_DSN_PUBLIC_KEY: &str = "31a5a894b4524f74a9a8d0e27e21ba91";
+
+#[derive(Default, Clone)]
+pub struct CapturedOutcomes {
+    inner: Arc<Mutex<Vec<Outcome>>>,
+}
+
+impl CapturedOutcomes {
+    pub fn is_empty(&self) -> bool {
+        self.inner.lock().unwrap().is_empty()
+    }
+
+    pub fn clear(&self) -> &Self {
+        self.inner.lock().unwrap().clear();
+        self
+    }
+
+    pub fn pop(&self) -> Option<Outcome> {
+        self.inner.lock().unwrap().pop()
+    }
+
+    pub fn extend(&self, outcomes: Vec<Outcome>) {
+        self.inner.lock().unwrap().extend(outcomes);
+    }
+
+    pub fn push(&self, outcome: Outcome) {
+        self.inner.lock().unwrap().push(outcome);
+    }
+
+    pub fn take_index(&self, idx: usize) -> Outcome {
+        self.inner.lock().unwrap().remove(idx)
+    }
+
+    pub fn assert_outcome_qty(&self, qty: usize) -> &Self {
+        assert_eq!(self.inner.lock().unwrap().len(), qty);
+        self
+    }
+
+    pub fn assert_all_outcome_reasons(&self, reason: &str) -> &Self {
+        let guard = self.inner.lock().unwrap();
+
+        for outcome in guard.iter() {
+            assert_eq!(outcome.reason(), reason);
+        }
+        self
+    }
+
+    pub fn assert_all_outcome_id(&self, outcome_id: u64) -> &Self {
+        let guard = self.inner.lock().unwrap();
+
+        for outcome in guard.iter() {
+            assert_eq!(outcome.outcome(), outcome_id);
+        }
+        self
+    }
+
+    pub fn wait_for_outcome(&self, timeout: u64) -> &Self {
+        for _ in 0..timeout {
+            if !self.is_empty() {
+                return self;
+            }
+            std::thread::sleep(Duration::from_secs(1));
```

Comment:
Ideally, we don't wait a full second if the next item has already arrived. You could use condition variables for this, but there's probably a more idiomatic way to do it, e.g. channels or watchers. https://doc.rust-lang.org/std/sync/struct.Condvar.html

Reply:
thanks! will look into this